File: python/hermetix/brokers/kiwoom.py
"""키움증권 REST 모의투자 어댑터 (KRX 국내주식).

실측 기반 (mockapi.kiwoom.com, 2026-08):
- 모든 호출은 POST + api-id 헤더(TR)로 라우팅
- 응답 엔벨로프: return_code(0=성공) / return_msg
- 가격 필드에 등락 방향 부호가 붙는다 (cur_prc "-239500" = 하락 중인 239,500원) -> 절대값
- 금액 필드는 zero-padded 문자열 ("000000100000000" = 1억)
- TR당 초당 1회 유량 제한 -> 1.1s 쓰로틀 + 백오프 재시도
- 캔들은 일봉(DAY_1)만 지원 / 모의투자는 KRX 만
"""
from __future__ import annotations

import threading
import time
from datetime import datetime, timezone
from decimal import Decimal

from ..broker import KST, BrokerClient, Throttle, _Http, krx_calendar, krx_tick_round
from ..errors import AuthError, BrokerApiError, MarketClosedError, OrderNotFoundError, RateLimitError
from ..models import (
    Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
    Fill, Holding, MarketDay, Order, OrderSide, OrderStatus, OrderType, Quote,
)


def _signed(value, default: str = "0") -> Decimal:
    """등락 부호 접두 필드 ("-239500", "+1200") - 부호 유지 파싱."""
    text = str(value).strip() if value not in (None, "") else default
    return Decimal(text.lstrip("+") or default)


def _signed_or_none(value) -> Decimal | None:
    if value in (None, ""):
        return None
    try:
        return Decimal(str(value).strip().lstrip("+"))
    except Exception:
        return None


def _padded(value, default: str = "0") -> Decimal:
    """zero-padded 금액 ("000000100000000") 파싱."""
    text = str(value).strip() if value not in (None, "") else default
    return Decimal(text or default)


class KiwoomClient(BrokerClient):

    capabilities = BrokerCapabilities(
        broker_id="kiwoom",
        market="KRX",
        currency="KRW",
        candle_intervals=frozenset({CandleInterval.DAY_1}),
        client_order_id=False,
        native_bracket=False,
        fractional_shares=False,
        server_open_orders=True,  # ka10075 미체결 조회 제공
    )

    def __init__(self, appkey: str, secretkey: str,
                 base_url: str = "https://mockapi.kiwoom.com", throttle_seconds: float = 1.1):
        self._appkey = appkey
        self._secretkey = secretkey
        self._http = _Http(base_url)
        self._throttle = Throttle(throttle_seconds)
        self._token: str | None = None
        self._token_expires_at = 0.0
        self._token_lock = threading.Lock()

    # ------------------------------------------------------------------ market

    def get_quotes(self, symbols: list[str]) -> list[Quote]:
        quotes = []
        for symbol in symbols:
            node = self._call("/api/dostk/stkinfo", "ka10001", {"stk_cd": symbol})
            rate = _signed_or_none(node.get("flu_rt"))
            quotes.append(Quote(
                symbol=symbol,
                price=abs(_signed(node.get("cur_prc"))),
                bid_price=None, ask_price=None,
                volume=int(abs(_signed(node.get("trde_qty")))),
                change=_signed_or_none(node.get("pred_pre")),
                change_rate=(rate / 100) if rate is not None else None,  # % -> 비율
                timestamp=datetime.now(timezone.utc),
            ))
        return quotes

    def get_candles(self, symbol: str, interval: CandleInterval, limit: int | None = None) -> list[Candle]:
        if interval != CandleInterval.DAY_1:
            raise ValueError("키움 어댑터는 일봉(DAY_1)만 지원합니다")
        rows = self._call("/api/dostk/chart", "ka10081",
                          {"stk_cd": symbol, "base_dt": datetime.now(KST).strftime("%Y%m%d"),
                           "upd_stkpc_tp": "1"}).get("stk_dt_pole_chart_qry", [])
        candles = [
            Candle(
                timestamp=datetime.strptime(r["dt"], "%Y%m%d").replace(tzinfo=KST),
                open=abs(_signed(r.get("open_pric"))), high=abs(_signed(r.get("high_pric"))),
                low=abs(_signed(r.get("low_pric"))), close=abs(_signed(r.get("cur_prc"))),
                volume=int(abs(_signed(r.get("trde_qty")))),
            )
            for r in rows if r.get("dt")
        ]
        candles.sort(key=lambda c: c.timestamp)  # 키움 최신순 -> 과거→최신
        return candles[-limit:] if limit else candles

    def get_calendar(self) -> list[MarketDay]:
        return krx_calendar()

    # ----------------------------------------------------------------- account

    def get_account(self) -> Account:
        deposit = self._call("/api/dostk/acnt", "kt00001", {"qry_tp": "3"})
        balance = self._balance()
        cash = _padded(deposit.get("entr"))
        portfolio = _padded(balance.get("prsm_dpst_aset_amt"))
        if portfolio <= 0:
            portfolio = cash + _padded(balance.get("tot_evlt_amt"))
        return Account(account_id="kiwoom-mock", currency="KRW", cash=cash, portfolio_value=portfolio)

    def get_holdings(self) -> list[Holding]:
        holdings = []
        for row in self._balance().get("acnt_evlt_remn_indv_tot") or []:
            qty = _padded(row.get("rmnd_qty"))
            if qty <= 0:
                continue
            rate = _signed_or_none(row.get("prft_rt"))
            holdings.append(Holding(
                symbol=str(row.get("stk_cd", "")).removeprefix("A"),
                quantity=qty,
                avg_entry_price=_padded(row.get("pur_pric")),
                current_price=abs(_signed(row.get("cur_prc"))) if row.get("cur_prc") else None,
                market_value=_padded(row.get("evlt_amt")) if row.get("evlt_amt") else None,
                unrealized_pnl=_signed_or_none(row.get("evltv_prft")),
                unrealized_pnl_rate=(rate / 100) if rate is not None else None,
            ))
        return holdings

    def get_buying_power(self) -> Decimal:
        deposit = self._call("/api/dostk/acnt", "kt00001", {"qry_tp": "3"})
        power = _padded(deposit.get("ord_alow_amt"))
        return power if power > 0 else _padded(deposit.get("entr"))

    # ------------------------------------------------------------------ orders

    def create_order(self, request: CreateOrderRequest) -> Order:
        api_id = "kt10000" if request.side == OrderSide.BUY else "kt10001"
        is_limit = request.order_type == OrderType.LIMIT
        node = self._call("/api/dostk/ordr", api_id, {
            "dmst_stex_tp": "KRX",
            "stk_cd": request.symbol,
            "ord_qty": str(request.quantity),
            "ord_uv": str(krx_tick_round(request.limit_price)) if is_limit else "",
            "trde_tp": "0" if is_limit else "3",
            "cond_uv": "",
        })
        return Order(
            order_id=str(node.get("ord_no", "")),
            status=OrderStatus.SUBMITTED,
            symbol=request.symbol, side=request.side, order_type=request.order_type,
            quantity=request.quantity, limit_price=request.limit_price,
            filled_quantity=Decimal(0), submitted_at=datetime.now(timezone.utc),
        )

    def get_orders(self) -> list[Order]:
        return [self._open_order(row) for row in self._open_rows()]

    def get_order(self, order_id: str) -> Order:
        for row in self._open_rows():
            if str(row.get("ord_no", "")).lstrip("0") == order_id.lstrip("0"):
                return self._open_order(row)
        for row in self._fill_rows():
            if str(row.get("ord_no", "")).lstrip("0") == order_id.lstrip("0"):
                return Order(
                    order_id=order_id, status=OrderStatus.FILLED,
                    symbol=str(row.get("stk_cd", "")).removeprefix("A"),
                    filled_quantity=_padded(row.get("cntr_qty")),
                    avg_fill_price=abs(_signed(row.get("cntr_pric"))) if row.get("cntr_pric") else None,
                )
        return Order(order_id=order_id, status=OrderStatus.CANCELED)

    def cancel_order(self, order_id: str) -> Order:
        row = next((r for r in self._open_rows()
                    if str(r.get("ord_no", "")).lstrip("0") == order_id.lstrip("0")), None)
        if row is None:
            raise OrderNotFoundError("order-not-found", f"키움 미체결 주문을 찾을 수 없습니다: {order_id}")
        self._call("/api/dostk/ordr", "kt10003", {
            "dmst_stex_tp": "KRX", "orig_ord_no": order_id,
            "stk_cd": str(row.get("stk_cd", "")).removeprefix("A"), "cncl_qty": "0",
        })
        return Order(order_id=order_id, status=OrderStatus.CANCELED, canceled_at=datetime.now(timezone.utc))

    def get_fills(self) -> list[Fill]:
        return [
            Fill(
                fill_id=str(r.get("ord_no", "")), order_id=str(r.get("ord_no", "")),
                symbol=str(r.get("stk_cd", "")).removeprefix("A"),
                side=OrderSide.BUY if "매수" in str(r.get("io_tp_nm", "")) else OrderSide.SELL,
                quantity=_padded(r.get("cntr_qty")),
                price=abs(_signed(r.get("cntr_pric"))) if r.get("cntr_pric") else None,
            )
            for r in self._fill_rows()
        ]

    # ---------------------------------------------------------------- internal

    @staticmethod
    def _open_order(row: dict) -> Order:
        ord_qty = _padded(row.get("ord_qty"))
        remaining = _padded(row.get("oso_qty")) if row.get("oso_qty") else ord_qty
        filled = ord_qty - remaining
        return Order(
            order_id=str(row.get("ord_no", "")),
            status=OrderStatus.PARTIALLY_FILLED if filled > 0 else OrderStatus.SUBMITTED,
            symbol=str(row.get("stk_cd", "")).removeprefix("A"),
            side=OrderSide.BUY if "매수" in str(row.get("io_tp_nm", "")) else OrderSide.SELL,
            order_type=OrderType.LIMIT,
            quantity=ord_qty,
            limit_price=abs(_signed(row.get("ord_pric"))) if row.get("ord_pric") else None,
            filled_quantity=filled,
        )

    def _open_rows(self) -> list[dict]:
        return self._call("/api/dostk/acnt", "ka10075",
                          {"all_stk_tp": "0", "trde_tp": "0", "stk_cd": "", "stex_tp": "0"}).get("oso", [])

    def _fill_rows(self) -> list[dict]:
        return self._call("/api/dostk/acnt", "ka10076",
                          {"stk_cd": "", "qry_tp": "0", "sell_tp": "0", "ord_no": "", "stex_tp": "0"}).get("cntr", [])

    def _balance(self) -> dict:
        return self._call("/api/dostk/acnt", "kt00018", {"qry_tp": "1", "dmst_stex_tp": "KRX"})

    def _call(self, path: str, api_id: str, body: dict) -> dict:
        for attempt in range(4):
            try:
                return self._call_once(path, api_id, body)
            except RateLimitError:
                if attempt >= 3:
                    raise
                time.sleep(1.1 * (attempt + 1))  # TR당 유량 제한 백오프
        raise AssertionError("unreachable")

    def _call_once(self, path: str, api_id: str, body: dict) -> dict:
        self._throttle.wait()
        status, parsed = self._http.request(
            "POST", path,
            headers={"authorization": f"Bearer {self._get_token()}", "api-id": api_id},
            json_body=body)

        if status < 200 or status >= 300 or parsed.get("return_code") != 0:
            code = str(parsed.get("return_code"))
            msg = f"키움({api_id}) {parsed.get('return_msg', '')}".strip()
            if "요청 개수를 초과" in msg:
                raise RateLimitError(status, code, msg)
            if "장종료" in msg or "RC4058" in msg:
                raise MarketClosedError(status, code, msg)
            if status == 401:
                raise AuthError(status, code, msg)
            raise BrokerApiError(status, code, msg)
        return parsed

    def _get_token(self) -> str:
        if self._token and time.time() < self._token_expires_at - 300:
            return self._token
        with self._token_lock:
            if self._token and time.time() < self._token_expires_at - 300:
                return self._token
            self._throttle.wait()
            status, body = self._http.request(
                "POST", "/oauth2/token",
                json_body={"grant_type": "client_credentials",
                           "appkey": self._appkey, "secretkey": self._secretkey})
            if status != 200 or body.get("return_code") != 0 or not body.get("token"):
                raise AuthError(status, str(body.get("return_code")),
                                f"키움 토큰 발급 실패: {body.get('return_msg', '')}")
            self._token = body["token"]
            # expires_dt: yyyyMMddHHmmss (KST)
            try:
                expires = datetime.strptime(body["expires_dt"], "%Y%m%d%H%M%S").replace(tzinfo=KST)
                self._token_expires_at = expires.timestamp()
            except (KeyError, ValueError):
                self._token_expires_at = time.time() + 86400
            return self._token

File: js/src/brokers/kiwoom.ts
/**
 * 키움증권 REST 모의투자 어댑터 (KRX). 실측 기반 (2026-08).
 *
 * - POST + api-id 헤더 라우팅 / return_code(0=성공)
 * - 가격에 등락 부호 접두 (cur_prc "-239500") -> 절대값 / 금액은 zero-padded
 * - TR당 초당 1회 유량 제한 -> 1.1s 쓰로틀 + 백오프
 */
import { Decimal } from "decimal.js";
import {
  BrokerClient, D, DorNull, RateLimiter, httpJson, krxCalendar, krxTickRound, kstYyyymmdd,
} from "../broker.js";
import { AuthError, BrokerApiError, MarketClosedError, OrderNotFoundError, RateLimitError } from "../errors.js";
import type {
  Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
  Fill, Holding, MarketDay, Order, Quote,
  TradingEnvironment,
} from "../models.js";
import { symbolCodeFor } from "../models.js";

/** 등락 부호 접두 필드 파싱 - 부호 유지. */
const signed = (v: unknown, fallback = "0"): Decimal => {
  const text = v === null || v === undefined || v === "" ? fallback : String(v).trim().replace(/^\+/, "");
  return new Decimal(text || fallback);
};
const signedOrNull = (v: unknown): Decimal | null => {
  if (v === null || v === undefined || v === "") return null;
  try { return new Decimal(String(v).trim().replace(/^\+/, "")); } catch { return null; }
};

export class KiwoomClient implements BrokerClient {
  readonly capabilities: BrokerCapabilities = {
    brokerId: "kiwoom",
    market: "KRX",
    currency: "KRW",
    candleIntervals: new Set<CandleInterval>(["1d"]),
    clientOrderId: false,
    nativeBracket: false,
    fractionalShares: false,
    serverOpenOrders: true, // ka10075 미체결 조회 제공
    environments: new Set<TradingEnvironment>(["PAPER", "LIVE"]),
  };

  private token: string | null = null;
  private tokenExpiresAt = 0;
  /** TR 당 초당 1회 유량 제한 — 쓰로틀 + 백오프 재시도 */
  private readonly limiter: RateLimiter;

  static readonly PAPER_URL = "https://mockapi.kiwoom.com";
  static readonly LIVE_URL = "https://api.kiwoom.com";
  readonly baseUrl: string;

  /** baseUrl 을 비우면 환경에 따라 결정(모의 mockapi / 실전 api.kiwoom.com). TR ID 는 공통 */
  constructor(
    private readonly appkey: string,
    private readonly secretkey: string,
    baseUrl: string = "",
    throttleMs = 1100,
    readonly environment: TradingEnvironment = "PAPER",
  ) {
    this.baseUrl = baseUrl || (environment === "LIVE" ? KiwoomClient.LIVE_URL : KiwoomClient.PAPER_URL);
    this.limiter = new RateLimiter(throttleMs, 3, (attempt) => 1100 * attempt);
  }

  // ---------------------------------------------------------------- market

  async getQuotes(symbols: string[]): Promise<Quote[]> {
    const quotes: Quote[] = [];
    for (const symbol of symbols) {
      const node = await this.call("/api/dostk/stkinfo", "ka10001", { stk_cd: symbolCodeFor(this.capabilities, symbol) });
      const rate = signedOrNull(node.flu_rt);
      quotes.push({
        symbol,
        price: signed(node.cur_prc).abs(),
        bidPrice: null, askPrice: null,
        volume: Number(signed(node.trde_qty).abs()),
        change: signedOrNull(node.pred_pre),
        changeRate: rate ? rate.div(100) : null,
        timestamp: new Date(),
      });
    }
    return quotes;
  }

  async getCandles(symbol: string, interval: CandleInterval, limit?: number): Promise<Candle[]> {
    if (interval !== "1d") throw new Error("키움 어댑터는 일봉(1d)만 지원합니다");
    const body = await this.call("/api/dostk/chart", "ka10081",
      { stk_cd: symbolCodeFor(this.capabilities, symbol), base_dt: kstYyyymmdd(), upd_stkpc_tp: "1" });
    const candles = ((body.stk_dt_pole_chart_qry ?? []) as Record<string, unknown>[])
      .filter((r) => r.dt)
      .map((r) => ({
        timestamp: new Date(`${String(r.dt).slice(0, 4)}-${String(r.dt).slice(4, 6)}-${String(r.dt).slice(6, 8)}T00:00:00+09:00`),
        open: signed(r.open_pric).abs(), high: signed(r.high_pric).abs(),
        low: signed(r.low_pric).abs(), close: signed(r.cur_prc).abs(),
        volume: Number(signed(r.trde_qty).abs()),
      }))
      .sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
    return limit ? candles.slice(-limit) : candles;
  }

  async getCalendar(): Promise<MarketDay[]> {
    return krxCalendar();
  }

  // --------------------------------------------------------------- account

  async getAccount(): Promise<Account> {
    const deposit = await this.call("/api/dostk/acnt", "kt00001", { qry_tp: "3" });
    const balance = await this.balance();
    const cash = D(deposit.entr);
    let portfolio = D(balance.prsm_dpst_aset_amt);
    if (portfolio.lte(0)) portfolio = cash.plus(D(balance.tot_evlt_amt));
    return { accountId: "kiwoom-mock", currency: "KRW", cash, portfolioValue: portfolio, status: "ACTIVE" };
  }

  async getHoldings(): Promise<Holding[]> {
    return (((await this.balance()).acnt_evlt_remn_indv_tot ?? []) as Record<string, unknown>[])
      .filter((row) => D(row.rmnd_qty).gt(0))
      .map((row) => {
        const rate = signedOrNull(row.prft_rt);
        return {
          symbol: String(row.stk_cd ?? "").replace(/^A/, ""),
          quantity: D(row.rmnd_qty),
          avgEntryPrice: D(row.pur_pric),
          currentPrice: row.cur_prc ? signed(row.cur_prc).abs() : null,
          marketValue: row.evlt_amt ? D(row.evlt_amt) : null,
          unrealizedPnl: signedOrNull(row.evltv_prft),
          unrealizedPnlRate: rate ? rate.div(100) : null,
        };
      });
  }

  async getBuyingPower(): Promise<Decimal> {
    const deposit = await this.call("/api/dostk/acnt", "kt00001", { qry_tp: "3" });
    const power = D(deposit.ord_alow_amt);
    return power.gt(0) ? power : D(deposit.entr);
  }

  // ---------------------------------------------------------------- orders

  async createOrder(request: CreateOrderRequest): Promise<Order> {
    const apiId = request.side === "BUY" ? "kt10000" : "kt10001";
    const isLimit = request.orderType === "LIMIT";
    const node = await this.call("/api/dostk/ordr", apiId, {
      dmst_stex_tp: "KRX",
      stk_cd: symbolCodeFor(this.capabilities, request.symbol),
      ord_qty: request.quantity.toString(),
      ord_uv: isLimit ? krxTickRound(request.limitPrice!).toString() : "",
      trde_tp: isLimit ? "0" : "3",
      cond_uv: "",
    });
    return {
      orderId: String(node.ord_no ?? ""),
      status: "SUBMITTED",
      symbol: symbolCodeFor(this.capabilities, request.symbol), side: request.side, orderType: request.orderType,
      quantity: request.quantity, limitPrice: request.limitPrice ?? null,
      filledQuantity: new Decimal(0), submittedAt: new Date(),
    };
  }

  async getOrders(): Promise<Order[]> {
    return (await this.openRows()).map((row) => this.toOpenOrder(row));
  }

  async getOrder(orderId: string): Promise<Order> {
    const key = orderId.replace(/^0+/, "");
    for (const row of await this.openRows()) {
      if (String(row.ord_no ?? "").replace(/^0+/, "") === key) return this.toOpenOrder(row);
    }
    for (const row of await this.fillRows()) {
      if (String(row.ord_no ?? "").replace(/^0+/, "") === key) {
        return {
          orderId, status: "FILLED",
          symbol: String(row.stk_cd ?? "").replace(/^A/, ""),
          filledQuantity: D(row.cntr_qty),
          avgFillPrice: row.cntr_pric ? signed(row.cntr_pric).abs() : null,
        };
      }
    }
    return { orderId, status: "CANCELED" };
  }

  async cancelOrder(orderId: string): Promise<Order> {
    const key = orderId.replace(/^0+/, "");
    const row = (await this.openRows()).find((r) => String(r.ord_no ?? "").replace(/^0+/, "") === key);
    if (!row) throw new OrderNotFoundError("order-not-found", `키움 미체결 주문을 찾을 수 없습니다: ${orderId}`);
    await this.call("/api/dostk/ordr", "kt10003", {
      dmst_stex_tp: "KRX", orig_ord_no: orderId,
      stk_cd: String(row.stk_cd ?? "").replace(/^A/, ""), cncl_qty: "0",
    });
    return { orderId, status: "CANCELED", canceledAt: new Date() };
  }

  async getFills(): Promise<Fill[]> {
    return (await this.fillRows()).map((r) => ({
      fillId: String(r.ord_no ?? ""), orderId: String(r.ord_no ?? ""),
      symbol: String(r.stk_cd ?? "").replace(/^A/, ""),
      side: String(r.io_tp_nm ?? "").includes("매수") ? "BUY" : "SELL",
      quantity: D(r.cntr_qty),
      price: r.cntr_pric ? signed(r.cntr_pric).abs() : null,
    }));
  }

  // -------------------------------------------------------------- internal

  private toOpenOrder(row: Record<string, unknown>): Order {
    const ordQty = D(row.ord_qty);
    const remaining = row.oso_qty ? D(row.oso_qty) : ordQty;
    const filled = ordQty.minus(remaining);
    return {
      orderId: String(row.ord_no ?? ""),
      status: filled.gt(0) ? "PARTIALLY_FILLED" : "SUBMITTED",
      symbol: String(row.stk_cd ?? "").replace(/^A/, ""),
      side: String(row.io_tp_nm ?? "").includes("매수") ? "BUY" : "SELL",
      orderType: "LIMIT",
      quantity: ordQty,
      limitPrice: row.ord_pric ? signed(row.ord_pric).abs() : null,
      filledQuantity: filled,
    };
  }

  private async openRows(): Promise<Record<string, unknown>[]> {
    const body = await this.call("/api/dostk/acnt", "ka10075",
      { all_stk_tp: "0", trde_tp: "0", stk_cd: "", stex_tp: "0" });
    return (body.oso ?? []) as Record<string, unknown>[];
  }

  private async fillRows(): Promise<Record<string, unknown>[]> {
    const body = await this.call("/api/dostk/acnt", "ka10076",
      { stk_cd: "", qry_tp: "0", sell_tp: "0", ord_no: "", stex_tp: "0" });
    return (body.cntr ?? []) as Record<string, unknown>[];
  }

  private balance(): Promise<Record<string, unknown>> {
    return this.call("/api/dostk/acnt", "kt00018", { qry_tp: "1", dmst_stex_tp: "KRX" });
  }

  private async call(path: string, apiId: string, json: Record<string, string>): Promise<Record<string, unknown>> {
    return this.limiter.execute(() => this.callOnce(path, apiId, json), `키움 ${apiId}`);
  }

  private async callOnce(path: string, apiId: string, json: Record<string, string>): Promise<Record<string, unknown>> {
    const [status, body] = await httpJson(this.baseUrl + path, {
      method: "POST",
      headers: {
        "Content-Type": "application/json;charset=UTF-8",
        authorization: `Bearer ${await this.getToken()}`,
        "api-id": apiId,
      },
      body: JSON.stringify(json),
    });
    if (status < 200 || status >= 300 || body.return_code !== 0) {
      const code = String(body.return_code ?? "");
      const msg = `키움(${apiId}) ${body.return_msg ?? ""}`.trim();
      if (msg.includes("요청 개수를 초과")) throw new RateLimitError(status, code, msg);
      if (msg.includes("장종료") || msg.includes("RC4058")) throw new MarketClosedError(status, code, msg);
      if (status === 401) throw new AuthError(status, code, msg);
      throw new BrokerApiError(status, code, msg);
    }
    return body;
  }

  private async getToken(): Promise<string> {
    if (this.token && Date.now() < this.tokenExpiresAt - 300_000) return this.token;
    await this.limiter.throttle.wait();
    const [status, body] = await httpJson(`${this.baseUrl}/oauth2/token`, {
      method: "POST",
      headers: { "Content-Type": "application/json;charset=UTF-8" },
      body: JSON.stringify({ grant_type: "client_credentials", appkey: this.appkey, secretkey: this.secretkey }),
    });
    if (status !== 200 || body.return_code !== 0 || typeof body.token !== "string") {
      throw new AuthError(status, String(body.return_code ?? ""), `키움 토큰 발급 실패: ${body.return_msg ?? ""}`);
    }
    this.token = body.token;
    // expires_dt: yyyyMMddHHmmss (KST)
    const dt = String(body.expires_dt ?? "");
    this.tokenExpiresAt = dt.length === 14
      ? new Date(`${dt.slice(0, 4)}-${dt.slice(4, 6)}-${dt.slice(6, 8)}T${dt.slice(8, 10)}:${dt.slice(10, 12)}:${dt.slice(12, 14)}+09:00`).getTime()
      : Date.now() + 86400_000;
    return this.token;
  }
}
